File: sdk/typescript/src/local_rpc.ts
import { fromBinary, toBinary } from "@bufbuild/protobuf";
import { createConnection, type Socket } from "node:net";

import {
  ExchangeRequestSchema,
  ExchangeResponseSchema,
  type Envelope,
  type ExchangeRequest,
  type ExchangeResponse,
} from "../../../gen/typescript/nlos/sabi/v1/envelope_pb.ts";

const MAXIMUM_SCHEMA_FRAME_BYTES = 1024 * 1024;
const SUPPORTED_SCHEMA = "nlos.sabi.Envelope";
const SUPPORTED_MAJOR = 1;

export type IpcErrorCode =
  | "INVALID_CONFIG"
  | "CONNECT"
  | "READ"
  | "WRITE"
  | "TIMEOUT"
  | "FRAME_TOO_LARGE"
  | "COMPATIBILITY"
  | "BACKPRESSURE"
  | "CONNECTION_UNUSABLE"
  | "REQUEST_ID_MISMATCH";

export class IpcError extends Error {
  readonly code: IpcErrorCode;

  constructor(code: IpcErrorCode, message: string, options?: ErrorOptions) {
    super(message, options);
    this.name = "IpcError";
    this.code = code;
  }
}

export interface TransportConfig {
  maximumFrameBytes: number;
  connectTimeoutMs: number;
  readTimeoutMs: number;
  writeTimeoutMs: number;
}

const defaultConfig: TransportConfig = {
  maximumFrameBytes: MAXIMUM_SCHEMA_FRAME_BYTES,
  connectTimeoutMs: 5_000,
  readTimeoutMs: 5_000,
  writeTimeoutMs: 5_000,
};

export class LocalRpcClient {
  readonly #socket: Socket;
  readonly #config: TransportConfig;
  #inFlight = false;
  #usable = true;

  private constructor(socket: Socket, config: TransportConfig) {
    this.#socket = socket;
    this.#config = config;
    socket.on("error", () => {
      this.#usable = false;
    });
    socket.on("close", () => {
      this.#usable = false;
    });
  }

  static async connect(
    endpoint: string,
    overrides: Partial<TransportConfig> = {},
  ): Promise<LocalRpcClient> {
    if (endpoint.length === 0) {
      throw new IpcError("INVALID_CONFIG", "IPC endpoint must not be empty");
    }
    const config = validateConfig({ ...defaultConfig, ...overrides });
    const socket = createConnection({ path: endpoint });
    await waitForConnect(socket, config.connectTimeoutMs);
    return new LocalRpcClient(socket, config);
  }

  async exchange(request: ExchangeRequest): Promise<ExchangeResponse> {
    const requestEnvelope = requireCompatibleEnvelope(request.envelope);
    const wire = toBinary(ExchangeRequestSchema, request);
    ensureFrameBound(wire.byteLength, this.#config.maximumFrameBytes);
    if (this.#inFlight) {
      throw new IpcError(
        "BACKPRESSURE",
        "IPC client already has an in-flight call",
      );
    }
    if (!this.#usable) {
      throw new IpcError(
        "CONNECTION_UNUSABLE",
        "IPC connection is unusable; reconnect",
      );
    }

    this.#inFlight = true;
    try {
      const responsePromise = readFrame(
        this.#socket,
        this.#config.maximumFrameBytes,
        this.#config.readTimeoutMs,
      );
      try {
        await writeFrame(this.#socket, wire, this.#config.writeTimeoutMs);
      } catch (error) {
        void responsePromise.catch(() => undefined);
        throw error;
      }
      const responseWire = await responsePromise;
      const response = fromBinary(ExchangeResponseSchema, responseWire);
      const responseEnvelope = requireCompatibleEnvelope(response.envelope);
      if (!equalBytes(responseEnvelope.requestId, requestEnvelope.requestId)) {
        throw new IpcError(
          "REQUEST_ID_MISMATCH",
          "IPC response request_id does not match the request",
        );
      }
      return response;
    } catch (error) {
      this.#poison();
      if (error instanceof IpcError) {
        throw error;
      }
      throw new IpcError("READ", "IPC exchange failed", { cause: error });
    } finally {
      this.#inFlight = false;
    }
  }

  close(): void {
    this.#poison();
  }

  #poison(): void {
    this.#usable = false;
    this.#socket.destroy();
  }
}

function validateConfig(config: TransportConfig): TransportConfig {
  if (
    !Number.isSafeInteger(config.maximumFrameBytes) ||
    config.maximumFrameBytes <= 0 ||
    config.maximumFrameBytes > MAXIMUM_SCHEMA_FRAME_BYTES
  ) {
    throw new IpcError(
      "INVALID_CONFIG",
      "maximumFrameBytes must be within 1..=1048576",
    );
  }
  for (const [name, value] of [
    ["connectTimeoutMs", config.connectTimeoutMs],
    ["readTimeoutMs", config.readTimeoutMs],
    ["writeTimeoutMs", config.writeTimeoutMs],
  ] as const) {
    if (!Number.isFinite(value) || value <= 0) {
      throw new IpcError("INVALID_CONFIG", `${name} must be positive`);
    }
  }
  return Object.freeze({ ...config });
}

function waitForConnect(socket: Socket, timeoutMs: number): Promise<void> {
  return new Promise((resolve, reject) => {
    const timer = setTimeout(() => {
      finish(
        new IpcError("TIMEOUT", `IPC connect timed out after ${timeoutMs} ms`),
      );
    }, timeoutMs);
    const onConnect = (): void => finish();
    const onError = (cause: Error): void =>
      finish(new IpcError("CONNECT", "IPC connect failed", { cause }));
    const finish = (error?: IpcError): void => {
      clearTimeout(timer);
      socket.off("connect", onConnect);
      socket.off("error", onError);
      if (error === undefined) {
        resolve();
      } else {
        socket.destroy();
        reject(error);
      }
    };
    socket.once("connect", onConnect);
    socket.once("error", onError);
  });
}

function writeFrame(
  socket: Socket,
  wire: Uint8Array,
  timeoutMs: number,
): Promise<void> {
  const prefix = Buffer.allocUnsafe(4);
  prefix.writeUInt32BE(wire.byteLength);
  const frame = Buffer.concat([prefix, Buffer.from(wire)]);
  return new Promise((resolve, reject) => {
    const timer = setTimeout(
      () => reject(new IpcError("TIMEOUT", "IPC write timed out")),
      timeoutMs,
    );
    socket.write(frame, (cause?: Error | null) => {
      clearTimeout(timer);
      if (cause == null) {
        resolve();
      } else {
        reject(new IpcError("WRITE", "IPC write failed", { cause }));
      }
    });
  });
}

function readFrame(
  socket: Socket,
  maximumFrameBytes: number,
  timeoutMs: number,
): Promise<Uint8Array> {
  return new Promise((resolve, reject) => {
    let buffered = Buffer.alloc(0);
    let declared: number | undefined;
    const timer = setTimeout(
      () => finish(new IpcError("TIMEOUT", "IPC read timed out")),
      timeoutMs,
    );
    const onData = (chunk: Buffer): void => {
      if (buffered.byteLength + chunk.byteLength > maximumFrameBytes + 4) {
        finish(
          new IpcError(
            "FRAME_TOO_LARGE",
            "IPC peer sent more bytes than the configured unary frame bound",
          ),
        );
        return;
      }
      buffered = Buffer.concat([buffered, chunk]);
      if (declared === undefined && buffered.byteLength >= 4) {
        declared = buffered.readUInt32BE(0);
        try {
          ensureFrameBound(declared, maximumFrameBytes);
        } catch (error) {
          finish(error as IpcError);
          return;
        }
      }
      if (declared !== undefined && buffered.byteLength >= declared + 4) {
        if (buffered.byteLength !== declared + 4) {
          finish(
            new IpcError(
              "READ",
              "IPC peer sent trailing bytes after a unary response frame",
            ),
          );
          return;
        }
        finish(undefined, buffered.subarray(4));
      }
    };
    const onError = (cause: Error): void =>
      finish(new IpcError("READ", "IPC read failed", { cause }));
    const onClose = (): void =>
      finish(new IpcError("READ", "IPC peer disconnected mid-exchange"));
    const finish = (error?: IpcError, wire?: Uint8Array): void => {
      clearTimeout(timer);
      socket.off("data", onData);
      socket.off("error", onError);
      socket.off("close", onClose);
      if (error === undefined && wire !== undefined) {
        resolve(wire);
      } else {
        reject(error ?? new IpcError("READ", "IPC response is missing"));
      }
    };
    socket.on("data", onData);
    socket.once("error", onError);
    socket.once("close", onClose);
  });
}

function ensureFrameBound(actual: number, maximum: number): void {
  if (actual > maximum) {
    throw new IpcError(
      "FRAME_TOO_LARGE",
      `IPC frame has ${actual} bytes; maximum is ${maximum}`,
    );
  }
}

function requireCompatibleEnvelope(envelope: Envelope | undefined): Envelope {
  if (envelope === undefined || envelope.schema === undefined) {
    throw new IpcError("COMPATIBILITY", "schema identity is missing");
  }
  if (envelope.schema.name !== SUPPORTED_SCHEMA) {
    throw new IpcError(
      "COMPATIBILITY",
      `schema ${JSON.stringify(envelope.schema.name)} is not registered`,
    );
  }
  if (envelope.schema.major !== SUPPORTED_MAJOR) {
    throw new IpcError(
      "COMPATIBILITY",
      `unsupported schema major ${envelope.schema.major}`,
    );
  }
  if (envelope.schema.criticalExtensionIds.length !== 0) {
    throw new IpcError(
      "COMPATIBILITY",
      `unsupported critical extension ${envelope.schema.criticalExtensionIds[0]}`,
    );
  }
  if (envelope.requestId.byteLength !== 16) {
    throw new IpcError("COMPATIBILITY", "request_id must contain 16 bytes");
  }
  if (envelope.service.length === 0 || envelope.method.length === 0) {
    throw new IpcError(
      "COMPATIBILITY",
      "service and method must not be empty",
    );
  }
  return envelope;
}

function equalBytes(left: Uint8Array, right: Uint8Array): boolean {
  return (
    left.byteLength === right.byteLength &&
    left.every((value, index) => value === right[index])
  );
}

File: sdk/python/nlos_sdk/local_rpc.py
"""Bounded asyncio client for the transport-neutral NLOS local RPC PoC."""

from __future__ import annotations

import asyncio
import math
import sys
from dataclasses import dataclass
from typing import Any, Callable, Coroutine

from google.protobuf.message import DecodeError
from nlos.sabi.v1.envelope_pb2 import ExchangeRequest, ExchangeResponse

MAXIMUM_SCHEMA_FRAME_BYTES = 1024 * 1024
SUPPORTED_SCHEMA = "nlos.sabi.Envelope"
SUPPORTED_MAJOR = 1


class IpcError(Exception):
    """Typed local IPC failure."""

    def __init__(self, code: str, message: str) -> None:
        super().__init__(message)
        self.code = code


@dataclass(frozen=True, slots=True)
class TransportConfig:
    """Bounded local transport settings."""

    maximum_frame_bytes: int = MAXIMUM_SCHEMA_FRAME_BYTES
    connect_timeout: float = 5.0
    read_timeout: float = 5.0
    write_timeout: float = 5.0

    def __post_init__(self) -> None:
        if (
            isinstance(self.maximum_frame_bytes, bool)
            or not isinstance(self.maximum_frame_bytes, int)
            or not 0 < self.maximum_frame_bytes <= MAXIMUM_SCHEMA_FRAME_BYTES
        ):
            raise IpcError(
                "INVALID_CONFIG",
                "maximum_frame_bytes must be within 1..=1048576",
            )
        timeouts = (
            self.connect_timeout,
            self.read_timeout,
            self.write_timeout,
        )
        if any(
            isinstance(value, bool)
            or not isinstance(value, (int, float))
            or not math.isfinite(value)
            or value <= 0
            for value in timeouts
        ):
            raise IpcError("INVALID_CONFIG", "timeouts must be finite and positive")


class LocalRpcClient:
    """Single-connection, single-in-flight local RPC client."""

    def __init__(
        self,
        reader: asyncio.StreamReader,
        writer: asyncio.StreamWriter,
        config: TransportConfig,
    ) -> None:
        self._reader = reader
        self._writer = writer
        self._config = config
        self._in_flight = False
        self._usable = True

    @classmethod
    async def connect(
        cls,
        endpoint: str,
        config: TransportConfig | None = None,
    ) -> LocalRpcClient:
        """Connect to a ServiceDirectory-resolved local endpoint."""

        if not endpoint:
            raise IpcError("INVALID_CONFIG", "IPC endpoint must not be empty")
        resolved = config or TransportConfig()
        try:
            reader, writer = await asyncio.wait_for(
                _open_platform_stream(endpoint),
                timeout=resolved.connect_timeout,
            )
        except TimeoutError as error:
            raise IpcError("TIMEOUT", "IPC connect timed out") from error
        except OSError as error:
            raise IpcError("CONNECT", "IPC connect failed") from error
        return cls(reader, writer, resolved)

    async def exchange(self, request: ExchangeRequest) -> ExchangeResponse:
        """Execute one unary call with immediate concurrent backpressure."""

        request_envelope = _require_compatible_envelope(request)
        wire = request.SerializeToString()
        _ensure_frame_bound(len(wire), self._config.maximum_frame_bytes)
        if self._in_flight:
            raise IpcError("BACKPRESSURE", "IPC client already has an in-flight call")
        if not self._usable:
            raise IpcError("CONNECTION_UNUSABLE", "IPC connection is unusable; reconnect")

        self._in_flight = True
        operation = "WRITE"
        try:
            prefix = len(wire).to_bytes(4, byteorder="big")
            self._writer.write(prefix + wire)
            await asyncio.wait_for(
                self._writer.drain(),
                timeout=self._config.write_timeout,
            )
            operation = "READ"
            response_wire = await asyncio.wait_for(
                _read_frame(self._reader, self._config.maximum_frame_bytes),
                timeout=self._config.read_timeout,
            )
            response = ExchangeResponse()
            response.ParseFromString(response_wire)
            response_envelope = _require_compatible_envelope(response)
            if response_envelope.request_id != request_envelope.request_id:
                raise IpcError(
                    "REQUEST_ID_MISMATCH",
                    "IPC response request_id does not match the request",
                )
            return response
        except asyncio.IncompleteReadError as error:
            await self._poison()
            raise IpcError("READ", "IPC peer disconnected mid-frame") from error
        except TimeoutError as error:
            await self._poison()
            raise IpcError("TIMEOUT", "IPC exchange timed out") from error
        except DecodeError as error:
            await self._poison()
            raise IpcError("COMPATIBILITY", "malformed protobuf response") from error
        except IpcError:
            await self._poison()
            raise
        except Exception as error:
            await self._poison()
            raise IpcError(operation, f"IPC {operation.lower()} failed") from error
        finally:
            self._in_flight = False

    async def close(self) -> None:
        """Close the stream and make the client permanently unusable."""

        await self._poison()

    async def _poison(self) -> None:
        self._usable = False
        self._writer.close()
        try:
            await self._writer.wait_closed()
        except OSError:
            pass


async def _open_platform_stream(
    endpoint: str,
) -> tuple[asyncio.StreamReader, asyncio.StreamWriter]:
    if sys.platform != "win32":
        return await asyncio.open_unix_connection(endpoint)

    loop = asyncio.get_running_loop()
    create_pipe_connection = getattr(loop, "create_pipe_connection", None)
    if create_pipe_connection is None:
        raise IpcError(
            "UNSUPPORTED_PLATFORM",
            "the active Windows event loop cannot connect to named pipes",
        )
    reader = asyncio.StreamReader(limit=MAXIMUM_SCHEMA_FRAME_BYTES + 4)
    protocol = asyncio.StreamReaderProtocol(reader)
    factory: Callable[[], asyncio.StreamReaderProtocol] = lambda: protocol
    connect: Callable[
        [Callable[[], asyncio.StreamReaderProtocol], str],
        Coroutine[Any, Any, tuple[asyncio.BaseTransport, asyncio.BaseProtocol]],
    ] = create_pipe_connection
    transport, _ = await connect(factory, endpoint)
    writer = asyncio.StreamWriter(transport, protocol, reader, loop)
    return reader, writer


def _ensure_frame_bound(actual: int, maximum: int) -> None:
    if actual > maximum:
        raise IpcError(
            "FRAME_TOO_LARGE",
            f"IPC frame has {actual} bytes; maximum is {maximum}",
        )


async def _read_frame(reader: asyncio.StreamReader, maximum: int) -> bytes:
    prefix = await reader.readexactly(4)
    declared = int.from_bytes(prefix, byteorder="big")
    _ensure_frame_bound(declared, maximum)
    return await reader.readexactly(declared)


def _require_compatible_envelope(message: ExchangeRequest | ExchangeResponse) -> Any:
    if not message.HasField("envelope") or not message.envelope.HasField("schema"):
        raise IpcError("COMPATIBILITY", "schema identity is missing")
    envelope = message.envelope
    if envelope.schema.name != SUPPORTED_SCHEMA:
        raise IpcError(
            "COMPATIBILITY",
            f"schema {envelope.schema.name!r} is not registered",
        )
    if envelope.schema.major != SUPPORTED_MAJOR:
        raise IpcError(
            "COMPATIBILITY",
            f"unsupported schema major {envelope.schema.major}",
        )
    if envelope.schema.critical_extension_ids:
        raise IpcError(
            "COMPATIBILITY",
            "unsupported critical extension "
            f"{envelope.schema.critical_extension_ids[0]}",
        )
    if len(envelope.request_id) != 16:
        raise IpcError("COMPATIBILITY", "request_id must contain 16 bytes")
    if not envelope.service or not envelope.method:
        raise IpcError("COMPATIBILITY", "service and method must not be empty")
    return envelope
